lib/galaxy/model/__init__.py

```diff
@@ -6073,8 +6073,9 @@ def to_history_dataset_association(self, target_history, parent_id=None, add_to_
         hda.metadata = self.metadata
         if add_to_history and target_history:
             target_history.add_dataset(hda)
```

Comment:
add_dataset also causes a flush, the modern way of handling this is `history.stage_addition` and then at the end `history.add_pending_items`

With optimizations like this it's usually a good idea to profile the changes so you can see a before/after, https://docs.galaxyproject.org/en/master/dev/finding_and_improving_slow_code.html has some pointers on that,